From patchwork Wed Jun  5 10:09:09 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10976565
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CC11076
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:47 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id BDB1B1FF29
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:47 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B189E28911; Wed,  5 Jun 2019 10:09:47 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2039F1FF29
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:46 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727303AbfFEKJc (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Jun 2019 06:09:32 -0400
Received: from mail-pg1-f193.google.com ([209.85.215.193]:41763 "EHLO
        mail-pg1-f193.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727071AbfFEKJb (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Jun 2019 06:09:31 -0400
Received: by mail-pg1-f193.google.com with SMTP id 83so5306793pgg.8;
        Wed, 05 Jun 2019 03:09:31 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=ZZm7DGje/zlhjtpgaoTPoEdNdJYzoHqe7K5tKElHL8Y=;
        b=HDj23HZb4qw/LTHKStHh0a8pmk3wkgcZ3Cz/08O6TLTnSbfgsvBj6J1FdfnRi3mN1L
         L/AWVl1zcd32QUrHZzdtFRjk0NgCUHuefoDdBtuLRSCdynuajgLCxBfGLB6HcxOecTRZ
         pM7hnpwlpXwOQTlR36Gy6YoX/M+pHBcsvDZqlkYgXi4Uq/avv5BvHqSP7enukwPTImBH
         twY/c16H7mZzUQfBtGK4uD8wVc+tRVSIISZ/BlAs606vzgL0eShKJDyKshUtogS/MPVI
         6nCteP+uD7aI9uwJNg3KMjAqj8dLK62peCrY2vjA+28wQeUPK+R2UFqboO8PVd5c7U/K
         2Ifg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=ZZm7DGje/zlhjtpgaoTPoEdNdJYzoHqe7K5tKElHL8Y=;
        b=J73MjI0sHGBwzHOnYoH2rNUQSW+wlO6vxdlobXiwElHmxjL+pPjoXgZiYPayOoluAu
         Gk+ZoXsGl1MvHkW0Eiwh+5GpFNBd03w+Yj9igufbRXF1As7fH45ciRrmdTf+0RWXGYuk
         WtpbTel2M4aV2HN4iLBddraLesB3re+t33FATKlbiqf3BhxGUlORq+d03jiX7ZlO2+pW
         Jl2zhDOJvZZ+0SctC9/rp6oHPSxzbDNHxGK6Yb3JV7Glw62sX6WDjsosj+qEScahCCRZ
         k+IGhetxglk9akc0+K3CqY3lRA/NDNAYQGMCOSgQjTDP8Mt1qtlcLSnK+DuFOky/87Xj
         v+3g==
X-Gm-Message-State: APjAAAXzLaq5wgkfmK/HCG/NHn92mhSjDtcj+iTyH0/cIgiJk4r0SJlz
        negW2g5E6v+SmyBcuNFguJ4Z0797
X-Google-Smtp-Source: 
 APXvYqxPsHm/fbZhg0wAMb/lhSCMH63kF+BsJoqoQWW7FESWmJnaQjWnY0J9XD+tV2jENP7lSFOemA==
X-Received: by 2002:a63:b1d:: with SMTP id 29mr3256286pgl.103.1559729370327;
        Wed, 05 Jun 2019 03:09:30 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 v9sm19030010pfm.34.2019.06.05.03.09.28
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Wed, 05 Jun 2019 03:09:29 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>
Subject: [PATCH 1/3] KVM: LAPIC: Make lapic timer unpinned when timer is
 injected by posted-interrupt
Date: Wed,  5 Jun 2019 18:09:09 +0800
Message-Id: <1559729351-20244-2-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
References: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Make lapic timer unpinned when timer is injected by posted-interrupt,
the emulated timer can be offload to the housekeeping cpus.

This patch introduces a new kvm module parameter, it is false by default.
The host admin can enable it after fine tuned, e.g. dedicated instances 
scenario w/ nohz_full cover the pCPUs which vCPUs resident, several pCPUs 
surplus for housekeeping, disable mwait/hlt/pause vmexits to occupy the 
pCPUs, fortunately preemption timer is disabled after mwait is exposed 
to guest which makes emulated timer offload can be possible. 

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/include/asm/kvm_host.h |  1 +
 arch/x86/kvm/lapic.c            | 20 ++++++++++++++++----
 arch/x86/kvm/svm.c              |  5 +++++
 arch/x86/kvm/vmx/vmx.c          |  9 +++++++++
 4 files changed, 31 insertions(+), 4 deletions(-)

diff --git a/arch/x86/include/asm/kvm_host.h b/arch/x86/include/asm/kvm_host.h
index aeadbc7..ccb3d61 100644
--- a/arch/x86/include/asm/kvm_host.h
+++ b/arch/x86/include/asm/kvm_host.h
@@ -1090,6 +1090,7 @@ struct kvm_x86_ops {
 	void (*set_virtual_apic_mode)(struct kvm_vcpu *vcpu);
 	void (*set_apic_access_page_addr)(struct kvm_vcpu *vcpu, hpa_t hpa);
 	void (*deliver_posted_interrupt)(struct kvm_vcpu *vcpu, int vector);
+	bool (*pi_inject_timer_enabled)(struct kvm_vcpu *vcpu);
 	int (*sync_pir_to_irr)(struct kvm_vcpu *vcpu);
 	int (*set_tss_addr)(struct kvm *kvm, unsigned int addr);
 	int (*set_identity_map_addr)(struct kvm *kvm, u64 ident_addr);
diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index fcf42a3..8c9c14d 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -127,6 +127,12 @@ static inline u32 kvm_x2apic_id(struct kvm_lapic *apic)
 	return apic->vcpu->vcpu_id;
 }
 
+static inline bool posted_interrupt_inject_timer(struct kvm_vcpu *vcpu)
+{
+	return (kvm_x86_ops->pi_inject_timer_enabled(vcpu) &&
+		kvm_mwait_in_guest(vcpu->kvm));
+}
+
 static inline bool kvm_apic_map_get_logical_dest(struct kvm_apic_map *map,
 		u32 dest_id, struct kvm_lapic ***cluster, u16 *mask) {
 	switch (map->mode) {
@@ -1581,7 +1587,9 @@ static void start_sw_tscdeadline(struct kvm_lapic *apic)
 	    likely(ns > apic->lapic_timer.timer_advance_ns)) {
 		expire = ktime_add_ns(now, ns);
 		expire = ktime_sub_ns(expire, ktimer->timer_advance_ns);
-		hrtimer_start(&ktimer->timer, expire, HRTIMER_MODE_ABS_PINNED);
+		hrtimer_start(&ktimer->timer, expire,
+			posted_interrupt_inject_timer(vcpu) ?
+			HRTIMER_MODE_ABS : HRTIMER_MODE_ABS_PINNED);
 	} else
 		apic_timer_expired(apic);
 
@@ -1683,7 +1691,8 @@ static void start_sw_period(struct kvm_lapic *apic)
 
 	hrtimer_start(&apic->lapic_timer.timer,
 		apic->lapic_timer.target_expiration,
-		HRTIMER_MODE_ABS_PINNED);
+		posted_interrupt_inject_timer(apic->vcpu) ?
+		HRTIMER_MODE_ABS : HRTIMER_MODE_ABS_PINNED);
 }
 
 bool kvm_lapic_hv_timer_in_use(struct kvm_vcpu *vcpu)
@@ -2320,7 +2329,8 @@ int kvm_create_lapic(struct kvm_vcpu *vcpu, int timer_advance_ns)
 	apic->vcpu = vcpu;
 
 	hrtimer_init(&apic->lapic_timer.timer, CLOCK_MONOTONIC,
-		     HRTIMER_MODE_ABS_PINNED);
+		posted_interrupt_inject_timer(vcpu) ?
+		HRTIMER_MODE_ABS : HRTIMER_MODE_ABS_PINNED);
 	apic->lapic_timer.timer.function = apic_timer_fn;
 	if (timer_advance_ns == -1) {
 		apic->lapic_timer.timer_advance_ns = 1000;
@@ -2509,7 +2519,9 @@ void __kvm_migrate_apic_timer(struct kvm_vcpu *vcpu)
 
 	timer = &vcpu->arch.apic->lapic_timer.timer;
 	if (hrtimer_cancel(timer))
-		hrtimer_start_expires(timer, HRTIMER_MODE_ABS_PINNED);
+		hrtimer_start_expires(timer,
+			posted_interrupt_inject_timer(vcpu) ?
+			HRTIMER_MODE_ABS : HRTIMER_MODE_ABS_PINNED);
 }
 
 /*
diff --git a/arch/x86/kvm/svm.c b/arch/x86/kvm/svm.c
index 302cb40..aee1d91 100644
--- a/arch/x86/kvm/svm.c
+++ b/arch/x86/kvm/svm.c
@@ -5173,6 +5173,11 @@ static void svm_deliver_avic_intr(struct kvm_vcpu *vcpu, int vec)
 		kvm_vcpu_wake_up(vcpu);
 }
 
+static inline bool svm_posted_interrupt_inject_timer(struct kvm_vcpu *vcpu)
+{
+	return false;
+}
+
 static void svm_ir_list_del(struct vcpu_svm *svm, struct amd_iommu_pi_data *pi)
 {
 	unsigned long flags;
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index da24f18..2b4fd61 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -101,6 +101,9 @@ module_param(fasteoi, bool, S_IRUGO);
 static bool __read_mostly enable_apicv = 1;
 module_param(enable_apicv, bool, S_IRUGO);
 
+bool __read_mostly pi_inject_timer = 0;
+module_param(pi_inject_timer, bool, S_IRUGO);
+
 /*
  * If nested=1, nested virtualization is supported, i.e., guests may use
  * VMX and be a hypervisor for its own guests. If nested=0, guests may not
@@ -3724,6 +3727,11 @@ static void vmx_deliver_posted_interrupt(struct kvm_vcpu *vcpu, int vector)
 		kvm_vcpu_kick(vcpu);
 }
 
+static bool vmx_pi_inject_timer_enabled(struct kvm_vcpu *vcpu)
+{
+	return pi_inject_timer && kvm_vcpu_apicv_active(vcpu);
+}
+
 /*
  * Set up the vmcs's constant host-state fields, i.e., host-state fields that
  * will not change in the lifetime of the guest.
@@ -7671,6 +7679,7 @@ static struct kvm_x86_ops vmx_x86_ops __ro_after_init = {
 	.guest_apic_has_interrupt = vmx_guest_apic_has_interrupt,
 	.sync_pir_to_irr = vmx_sync_pir_to_irr,
 	.deliver_posted_interrupt = vmx_deliver_posted_interrupt,
+	.pi_inject_timer_enabled = vmx_pi_inject_timer_enabled,
 
 	.set_tss_addr = vmx_set_tss_addr,
 	.set_identity_map_addr = vmx_set_identity_map_addr,

From patchwork Wed Jun  5 10:09:10 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10976563
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id C625176
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:46 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id B827D284B9
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:46 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id AA504288C4; Wed,  5 Jun 2019 10:09:46 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id CA512284F1
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:44 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727336AbfFEKJk (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Jun 2019 06:09:40 -0400
Received: from mail-pl1-f195.google.com ([209.85.214.195]:39175 "EHLO
        mail-pl1-f195.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727203AbfFEKJc (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Jun 2019 06:09:32 -0400
Received: by mail-pl1-f195.google.com with SMTP id g9so9512972plm.6;
        Wed, 05 Jun 2019 03:09:32 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=qkxXjj4yDg1YqzQ56xMNaH8WCEkd0pao4XID5lR33u0=;
        b=qq4KcFpg6V/NYUJmTiK0L3VbS952jUCLbUCBYxxjqzMweeyqhdQOXFElFN5S7+7tqJ
         SPbduEujYwjMkH4nr99Lh36WjKYgleWwZR7MqFL9sPZ0Cm3bX47+LoiY/19mRV6N9a8h
         NKfuw3wSPlkPb/SxvvoSIlKNETvT7rCzhO5jY/T4UI3b1oXdYKbe9eN98QyeQo1/O2BW
         vJGLyb+XAKbnrA/HxVehYxHiS5LNPGqny+laQicKqxwFUvANC2YwkT9YEIF5F3ZSYgI6
         eyFoWgFcc0GVO8xzRKli5A7RW+rjjSI8r+Z/DfJgwACJGbgSy245Fcp9dGUabbEvAAsd
         4iwQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=qkxXjj4yDg1YqzQ56xMNaH8WCEkd0pao4XID5lR33u0=;
        b=p1RcEw6h/3HvmYDWcSA3BqKNeUsKe6z2VLYEaSJQsKTjGV4B3vm+pOuVG83Xjo/dca
         6rr0MuBARTbL5DwxBCnZnyQ07Zh93D8ZgPobCZ3IiMMMG7c3oV1/gG7WUG9hvX0EcdmS
         WVBDrflrB0WNhqsNS+lTgsHbVDOl5r5/1FKL5DrcCMsp1UPAAQe6c8NWeMvLBMYCt8xU
         HilsSSpABInA1Mln3cYiF+KrySc8i4buUPKfu3tbl8b90JqjE0nMv/hXFr5k+y4hqjse
         KC3+Y8a13FMDeQPHyj+Up4xMRAs+5zWTKwcUgiHezueH6i2tyrf6qsFiRywkpiLuF7Mx
         7YTg==
X-Gm-Message-State: APjAAAVAIMiUeRyaLhpUD9EOzenUprqw4tLmZN4LiMLwQpFDbtdRXIEZ
        RVWBLky82wpsIgdvtG7fvz87GtJO
X-Google-Smtp-Source: 
 APXvYqxtALWInCOxeqKT/4oHumokJPlHOQVmMyB+FdMSWRr48iiVjvrZcg0Ag76eYJlpIMGO2v+dAw==
X-Received: by 2002:a17:902:2862:: with SMTP id
 e89mr42874547plb.258.1559729372086;
        Wed, 05 Jun 2019 03:09:32 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 v9sm19030010pfm.34.2019.06.05.03.09.30
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Wed, 05 Jun 2019 03:09:31 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>
Subject: [PATCH 2/3] KVM: LAPIC: lapic timer is injected by posted interrupt
Date: Wed,  5 Jun 2019 18:09:10 +0800
Message-Id: <1559729351-20244-3-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
References: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

Dedicated instances are currently disturbed by unnecessary jitter due 
to the emulated lapic timers fire on the same pCPUs which vCPUs resident.
There is no hardware virtual timer on Intel for guest like ARM. Both 
programming timer in guest and the emulated timer fires incur vmexits.
This patchset tries to avoid vmexit which is incurred by the emulated 
timer fires in dedicated instance scenario. 

When nohz_full is enabled in dedicated instances scenario, the unpinned 
timer will be moved to the nearest busy housekeepers after commit 444969223c8
("sched/nohz: Fix affine unpinned timers mess"). However, KVM always makes 
lapic timer pinned to the pCPU which vCPU residents, the reason is explained 
by commit 61abdbe0 (kvm: x86: make lapic hrtimer pinned). Actually, these 
emulated timers can be offload to the housekeeping cpus since APICv 
is really common in recent years. The guest timer interrupt is injected by 
posted-interrupt which is delivered by housekeeping cpu once the emulated 
timer fires. 

3%~5% redis performance benefit can be observed on Skylake server.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kvm/lapic.c | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index 8c9c14d..e9db086 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -1465,6 +1465,23 @@ static void apic_timer_expired(struct kvm_lapic *apic)
 }
 
 /*
+ * On APICv, lapic timer is injected by posted interrupt
+ * to dedicated instance.
+ */
+static void apic_timer_expired_pi(struct kvm_lapic *apic)
+{
+	struct kvm_timer *ktimer = &apic->lapic_timer;
+
+	kvm_apic_local_deliver(apic, APIC_LVTT);
+	if (apic_lvtt_tscdeadline(apic))
+		ktimer->tscdeadline = 0;
+	if (apic_lvtt_oneshot(apic)) {
+		ktimer->tscdeadline = 0;
+		ktimer->target_expiration = 0;
+	}
+}
+
+/*
  * On APICv, this test will cause a busy wait
  * during a higher-priority task.
  */
@@ -2297,7 +2314,10 @@ static enum hrtimer_restart apic_timer_fn(struct hrtimer *data)
 	struct kvm_timer *ktimer = container_of(data, struct kvm_timer, timer);
 	struct kvm_lapic *apic = container_of(ktimer, struct kvm_lapic, lapic_timer);
 
-	apic_timer_expired(apic);
+	if (unlikely(posted_interrupt_inject_timer(apic->vcpu)))
+		apic_timer_expired_pi(apic);
+	else
+		apic_timer_expired(apic);
 
 	if (lapic_is_periodic(apic)) {
 		advance_periodic_target_expiration(apic);

From patchwork Wed Jun  5 10:09:11 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 10976561
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id D55226C5
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:39 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id C5C831FF60
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:39 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id B9E66284F1; Wed,  5 Jun 2019 10:09:39 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 69856288C4
	for <patchwork-kvm@patchwork.kernel.org>;
 Wed,  5 Jun 2019 10:09:39 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727324AbfFEKJf (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Wed, 5 Jun 2019 06:09:35 -0400
Received: from mail-pl1-f196.google.com ([209.85.214.196]:45167 "EHLO
        mail-pl1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727112AbfFEKJe (ORCPT <rfc822;kvm@vger.kernel.org>);
        Wed, 5 Jun 2019 06:09:34 -0400
Received: by mail-pl1-f196.google.com with SMTP id x7so8527407plr.12;
        Wed, 05 Jun 2019 03:09:34 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references
         :mime-version:content-transfer-encoding;
        bh=dIuUcIyMK8nuPrLvAHMlgxxAYTbCV88vnbbR/Oftlkw=;
        b=mTFuorutizHMDSHbn3tP7SVn4HDvwoZqGcuzMbB/n4GQXN0kZmLz4uBbA3hI1+kTVc
         1Ln3yQC/A7lPuAme07EdEV1ZIrrOS+W/f6SNmQv4ttktnP9gc6vXwIjrXRpWBIiKrC2R
         cCRT9BtGKwbdmxcZM+r9+156WM10WG8fxOwKx8XehsgBSQrOjBYw2G6MnuIMzKBBqBBB
         cI83VBG5vYz0IM+MkNtYJTvvoP5+GH/nx7K+YwQGZjJXNefIPzEgPHbG4Cr7Dbl243wb
         F0UfgNcN3D/jo7e6/3sbWzkAsVaVWq0nFbs4UWnwnrKHUUkpOKe04GAzMYTxDX/Hbkbm
         pVEQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references:mime-version:content-transfer-encoding;
        bh=dIuUcIyMK8nuPrLvAHMlgxxAYTbCV88vnbbR/Oftlkw=;
        b=XMw4+KsVWBorQ57a7tFkIndO8bAuft61uxKFmsxNgWZpH/5GLGy+HVQLxY4GVLckoW
         Ab4+hy9dNKFVsc9TMd9L5Bx4Tf3nYqkuFlk+YRQ3X9JBlq12JqP+N5wYSgqv3rjIki8v
         rsazJboftWWkLIdS0vG7vOKyeSI+v9ALv62GHvar6vW5/p6Eig5Ue6DH1DKMg+HrEAEr
         LdswZ2CrAb+1zcYiLh6ewe0GXl2Cmt2St0Y93Qc8dsGbV4Ck07rIBJgLJUFBh8lYVYKl
         O4BlDNGQO4GfPRjnRS2LprfP9Iqeef5em9TFGMAjzByUL9w5Yg1UYfutLFWsbl1LX+H3
         xYtw==
X-Gm-Message-State: APjAAAVJ8chjn3w41AKjdvGbUIUzTVUPbfCK0vKBg3rr2x57aCEzvrU6
        Fcc0az9AUKVNOejg221+e3HsA5LS
X-Google-Smtp-Source: 
 APXvYqyBoWBc80zPeCCeF2ViNlmd8cdOjGvoi9xGqyokMwkW51ZaQFIT+cjnME8u/54BIHukQysgWQ==
X-Received: by 2002:a17:902:28c9:: with SMTP id
 f67mr42846474plb.19.1559729373861;
        Wed, 05 Jun 2019 03:09:33 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.123])
        by smtp.googlemail.com with ESMTPSA id
 v9sm19030010pfm.34.2019.06.05.03.09.32
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-SHA bits=128/128);
        Wed, 05 Jun 2019 03:09:33 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
 =?utf-8?b?UmFkaW0gS3LEjW3DocWZ?= <rkrcmar@redhat.com>
Subject: [PATCH 3/3] KVM: LAPIC: Ignore timer migration when lapic timer is
 injected by posted-interrupt
Date: Wed,  5 Jun 2019 18:09:11 +0800
Message-Id: <1559729351-20244-4-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
References: <1559729351-20244-1-git-send-email-wanpengli@tencent.com>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Wanpeng Li <wanpengli@tencent.com>

When lapic timer is injected by posted-interrupt, the emulated timer is 
offload to the housekeeping cpu. The timer interrupt will be delivered 
properly, no need to migrate timer.

Cc: Paolo Bonzini <pbonzini@redhat.com>
Cc: Radim Krčmář <rkrcmar@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kvm/lapic.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index e9db086..3bb4376 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -2534,7 +2534,8 @@ void __kvm_migrate_apic_timer(struct kvm_vcpu *vcpu)
 {
 	struct hrtimer *timer;
 
-	if (!lapic_in_kernel(vcpu))
+	if (!lapic_in_kernel(vcpu) ||
+		posted_interrupt_inject_timer(vcpu))
 		return;
 
 	timer = &vcpu->arch.apic->lapic_timer.timer;
